// dredd-transactions/dist/src/detect-transaction-example-numbers.js
'use strict';

// Provides index of requests and responses within given *transition*, sorted by
// their position in the original API Blueprint document (from first to last).
//
//     [
//         {
//             'type': 'httpResponse',
//             'transaction': fury.minim.elements.HttpTransaction()
//             'position': 85,
//         },
//         ...
//     ]
//
// ## Index Entry (object)
//
// - type: httpRequest, httpResponse (enum)
// - position (number) - Position of the first character relevant to
//   the request (or response) within the original API Blueprint document.
function createRequestsResponsesIndex(transitionElement) {
  return transitionElement.transactions.map(function (transactionElement) {
    var elements = [transactionElement.request, transactionElement.response];
    return elements.reduce(function (indexEntries, element) {
      if (element.sourceMapValue) {
        indexEntries.push({
          position: Math.max.apply(null, element.sourceMapValue.reduce(function (positions, current) {
            return positions.concat(current);
          }, [])),
          transactionElement,
          type: element.element
        });
      }
      return indexEntries;
    }, []);
  }).reduce(function (index, indexEntries) {
    return index.concat(indexEntries);
  }, []).sort(function (a, b) {
    return a.position - b.position;
  });
}

// Detects transaction example numbers for given transition element
//
// Returns an array of numbers, where indexes correspond to HTTP transactions
// within the transition and values represent the example numbers.
function detectTransactionExampleNumbers(transitionElement) {
  var requestsResponsesIndex = createRequestsResponsesIndex(transitionElement);
  var finalState = requestsResponsesIndex.reduce(function (previousState, indexEntry) {
    var currentState = Object.assign({}, previousState);

    switch (indexEntry.type) {
      case 'httpRequest':
        if (currentState.previousType === 'httpResponse') currentState.currentNo += 1;
        currentState.previousType = 'httpRequest';
        break;
      case 'httpResponse':
        currentState.previousType = 'httpResponse';
        break;
      default:
        return currentState;
    }

    currentState.transactionIndex.set(indexEntry.transactionElement, currentState.currentNo);
    return currentState;
  }, {
    currentNo: 1,
    transactionIndex: new Map(),
    previousType: 'httpRequest'
  });
  return Array.from(finalState.transactionIndex.values());
}

module.exports = detectTransactionExampleNumbers;
